packages/flutter/lib/src/services/asset_resolver.dart

```diff
@@ -0,0 +1,85 @@
+// Copyright 2016 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+import 'dart:async';
+import 'dart:collection';
+
+import 'asset_bundle.dart';
+
+/// Base class for resolver components that transform an input into part of
+/// a resolved asset's key.
+abstract class Qualifier {
+  String resolve();
+}
+
+/// Qualifier that knows how to find the nearest match among a list of
+/// candidates, based on a [SplayTreeMap] of normative values to candidates.
+class BreakpointQualifier extends Qualifier {
+  BreakpointQualifier({ this.candidates, this.value }) {
+    assert(this.candidates.length > 0);
+  }
+  final SplayTreeMap<double, String> candidates;
+  final double value;
+  String resolve() {
+    if (candidates.containsKey(value))
+      return candidates[value];
+    double lower = candidates.lastKeyBefore(value);
+    double upper = candidates.firstKeyAfter(value);
+    if (lower == null)
+      return candidates[upper];
+    if (upper == null)
+      return candidates[lower];
+    if (value > (lower + upper) / 2)
+      return candidates[upper];
+    else
+      return candidates[lower];
+  }
+}
+
+/// Base class for asset resolvers.
+abstract class AssetResolver {
+  /// Return a resolved asset key for the asset named [name].
+  Future<String> resolve(String name);
+}
+
+/// Base class for resolvers that fetch metadata from the asset bundle as an
+/// aid to resolution.
+abstract class MetadataAssetResolver extends AssetResolver {
+  MetadataAssetResolver({ this.bundle });
+  final AssetBundle bundle;
+  final Map<String, String> _metadataCache = new Map<String, String>();
+
+  Future<String> resolve(String name) async {
+    if (!_metadataCache.containsKey(name))
+      _metadataCache[name] = await bundle.loadString(constructKey(name));
+    return resolveFromMetadata(_metadataCache[name]);
+  }
+
+  /// Constructs the asset bundle key for fetching the metadata associated
+  /// with the asset with the given key. The default implementation simply
+  /// appends ".meta" to the key.
+  String constructKey(String name) {
+    return name + ".meta";
+  }
+
+  String resolveFromMetadata(String metadata);
+}
+
+/// Base class for resolvers that use a list of candidate keys as their
+/// metadata.
+///
+/// List entries are newline-separated in the metadata string.
+abstract class CandidateAssetResolver extends MetadataAssetResolver {
+
+  CandidateAssetResolver({ AssetBundle bundle }) : super(bundle: bundle);
+
+  String resolveFromMetadata(String metadata) {
+    List<String> candidates = metadata.split(new RegExp("\n+"));
+    if (candidates.last == '') // allow trailing newlines
+      candidates.removeLast();
```

Comment:
I'm torn about inventing yet another file format here...  Please at least store the RegExp in a private static so that we don't need to recompile the RegExp all the time.